src/proj.rs

```diff
@@ -0,0 +1,791 @@
+        );
+    let (
+        b11, b12, b13,
+        b21, b22, b23,
+        b31, b32, b33
+        ) = (
+        b[0], b[1], b[2],
+        b[3], b[4], b[5], 
+        b[6], b[7], b[8]
+        );
+
+    [a11*b11 + a12*b21 + a13*b31, a11*b12 + a12*b22 + a13*b32, a11*b13 + a12*b23 +a13*b33,
+     a21*b11 + a22*b21 + a23*b31, a21*b12 + a22*b22 + a23*b32, a21*b13 + a22*b23 + a23*b33,
+     a31*b11 + a32*b21 + a33*b31, a31*b12 + a32*b22 + a33*b32, a31*b13 + a32*b23 + a33*b33]
+}
+
+fn blend<P>(
+    top_left: P,
+    top_right: P,
+    bottom_left: P,
+    bottom_right: P,
+    right_weight: f32,
+    bottom_weight: f32,
+) -> P
+where
+    P: Pixel,
+    P::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let top = top_left.map2(&top_right, |u, v| {
+        P::Subpixel::clamp((1f32 - right_weight) * cast(u) + right_weight * cast(v))
+    });
+
+    let bottom = bottom_left.map2(&bottom_right, |u, v| {
+        P::Subpixel::clamp((1f32 - right_weight) * cast(u) + right_weight * cast(v))
+    });
+
+    top.map2(&bottom, |u, v| {
+        P::Subpixel::clamp((1f32 - bottom_weight) * cast(u) + bottom_weight * cast(v))
+    })
+}
+
+
+#[allow(deprecated)]
+fn interpolate<P>(image: &Image<P>, x: f32, y: f32, default: P) -> P
+where
+    P: Pixel + 'static,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let left = x.floor();
+    let right = left + 1f32;
+    let top = y.floor();
+    let bottom = top + 1f32;
+
+    let right_weight = x - left;
+    let bottom_weight = y - top;
+
+    // default if out of bound
+    let (width, height) = image.dimensions();
+    if left < 0f32 || right >= width as f32 || top < 0f32 || bottom >= height as f32 {
+        default
+    } else {
+        let (tl, tr, bl, br) = unsafe {
+            (
+                image.unsafe_get_pixel(left as u32, top as u32),
+                image.unsafe_get_pixel(right as u32, top as u32),
+                image.unsafe_get_pixel(left as u32, bottom as u32),
+                image.unsafe_get_pixel(right as u32, bottom as u32),
+            )
+        };
+        blend(tl, tr, bl, br, right_weight, bottom_weight)
+    }
+}
+
+#[inline(always)]
+#[allow(deprecated)]
+fn nearest<P: Pixel + 'static>(image: &Image<P>, x: f32, y: f32, default: P) -> P {
+    let rx = x.round() as u32;
+    let ry = y.round() as u32;
+    let (width, height) = image.dimensions();
+    if x.round() < 0f32 ||  y.round() < 0f32 {
+        default
+    } else if rx >= width || ry >= height {
+        default
+    } else {
+        unsafe { image.unsafe_get_pixel(rx, ry) }
+    }
+}
+
+/// How to handle pixels whose pre-image lies between input pixels.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum Interpolation {
+    /// Choose the nearest pixel to the pre-image of the
+    /// output pixel.
+    Nearest,
+    /// Bilinearly interpolate between the four pixels
+    /// closest to the pre-image of the output pixel.
+    Bilinear,
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```

Comment:
We should remove duplication between these tests and those in `affine`. If you merge the two files then we can at least put the equivalent tests for `some_specialised_function` and `equivalent_projective_transformation` next to each other.